### DISCUSSION TITLE
Difference between icu, snaphu, and snaphu2stage

### DISCUSSION THREAD

**Discussion by leeyhgeo** [isce-framework/isce2#700] [Q&A] [answered]
I'm using stackStripMap.py to process SAR images. I noticed there are icu, snaphu, and snaphu2stage options in --unw_method. I am confused about the difference between these unwrapping methods and suitable conditions, especially "snaphu" and "snaphu2stage."

**Comment by fdelgadodelapuente** [comment, marked as answer]
ICU is an improved version of the branch and cut algorithm from ROI_PAC. It is fast and works well if coherence is high and the data does not have a lot fringes.

SNAPHU_MCF is the Statistical Network Approach to PHase Unwrapping - minimum cost flow. It is slow but can unwrap well interferograms of low coherence and with many disconnected patches. You have to always manually check the quality of the unwrapping because it will unwrap even areas where there the coherence is zero.

**Reply by leeyhgeo** [reply]
Thank you for the clear explanation! 
What about the difference between normal SNAPHU and two-stage SNAPHU?

**Reply by fdelgadodelapuente** [reply]
snaphu2stage is for bridging an interferogram using the connected components mask. that file tells you which areas are self consistently unwrapped. You can also correct each connected components with an integer of 2pi by checking the phase differences with respect to adjacent patches. It can be tricky and tedious if you have too many connected components

**Reply by leeyhgeo** [reply]
I'm still confused about the step of "bridging an interferogram." 
Does it mean this step can improve the unwrapping results when discontinuities exist?
Is there any document or paper about the snaphu2stage method? I can hardly find information about it.

**Reply by fdelgadodelapuente** [reply]
Yes. It means that if you have two patches that belong to a connected component, but are separated by a clear phase discontinuity you can add /substract an integer of 2pi to one component and then level it with respect to the other one. You can repeat the process for as many components are required, and then reference the interferogram with respect to a non-deforming area. 

It can be easily implemented in a python script in which you load the unwrapped interferogram, the connected components mask, correct the interferograms and then export it to an ISCE RMG two band file.

Unfortunately there are not documents that explain all these technical details.
